File: test/harness/inventory.ts
import { createHash } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { HARNESS_DIR } from "./paths.js";
import { normalize } from "./normalize.js";
import { getProfile, PROFILES } from "./profiles.js";
import type { Manifest } from "./types.js";

export type InventoryCase = {
	id: string;
	profile: string;
	suite: "corpus" | "diagnostic";
	inputPath: string;
	goldenPath: string;
};
export function files(dir: string): string[] {
	if (!fs.existsSync(dir))
		throw new Error(`Missing required directory: ${dir}`);
	return fs
		.readdirSync(dir, { withFileTypes: true })
		.flatMap((e) =>
			e.isDirectory()
				? files(path.join(dir, e.name))
				: [path.join(dir, e.name)],
		)
		.sort();
}
export function json(file: string) {
	return JSON.parse(fs.readFileSync(file, "utf8"));
}
function safeId(id: string) {
	if (
		typeof id !== "string" ||
		!/^[a-zA-Z0-9_-]+(?:\/[a-zA-Z0-9_.-]+)*$/.test(id) ||
		id.split("/").includes("..")
	)
		throw new Error(`Invalid case ID: ${id}`);
}
/** Discover from inputs and assignments, never from available expected output. */
export function inventory(
	root = HARNESS_DIR,
	requireGoldens = true,
): InventoryCase[] {
	const manifest = json(path.join(root, "corpus/manifest.json")) as Manifest;
	if (!manifest.entries?.length || manifest.count !== manifest.entries.length)
		throw new Error("corpus manifest: empty or incorrect count");
	const entries: InventoryCase[] = [];
	const ids = new Set<string>();
	const counts: Record<string, number> = {};
	function add(
		suite: "corpus" | "diagnostic",
		id: string,
		inputPath: string,
		profiles: string[],
	) {
		safeId(id);
		if (ids.has(`${suite}/${id}`))
			throw new Error(
				`Duplicate ${suite} case ID: ${id} (${profiles.join(",")})`,
			);
		ids.add(`${suite}/${id}`);
		if (
			!Array.isArray(profiles) ||
			!profiles.length ||
			new Set(profiles).size !== profiles.length
		)
			throw new Error(`${id}: empty/duplicate profile assignments`);
		for (const profile of profiles) {
			try {
				getProfile(profile);
			} catch {
				throw new Error(`${id}: unknown profile ${profile}`);
			}
			const goldenPath = path.join(
				root,
				suite === "corpus" ? "goldens" : "diagnostics-goldens",
				profile,
				`${id}.html`,
			);
			if (!fs.existsSync(inputPath))
				throw new Error(`${profile} :: ${id}: missing input ${inputPath}`);
			if (requireGoldens && !fs.existsSync(goldenPath))
				throw new Error(`${profile} :: ${id}: missing golden ${goldenPath}`);
			entries.push({ suite, id, profile, inputPath, goldenPath });
		}
	}
	for (const entry of manifest.entries) {
		safeId(entry.id);
		if (entry.path !== `${entry.id}.md`)
			throw new Error(
				`${entry.id} (${entry.profiles}): input path differs from ID`,
			);
		const inputPath = path.join(root, "corpus/inputs", entry.path);
		const input = fs.readFileSync(inputPath, "utf8");
		if (
			input !== normalize(input) ||
			Buffer.byteLength(input) !== entry.bytes ||
			createHash("sha256").update(input).digest("hex") !== entry.sha256
		)
			throw new Error(
				`${entry.profiles.join(",")} :: ${entry.id}: input normalization, bytes or sha256 mismatch`,
			);
		counts[entry.source] = (counts[entry.source] ?? 0) + 1;
		add("corpus", entry.id, inputPath, entry.profiles);
	}
	if (
		JSON.stringify(Object.entries(counts).sort()) !==
		JSON.stringify(Object.entries(manifest.bySource).sort())
	)
		throw new Error("corpus manifest: bySource counts differ");
	const map = json(path.join(root, "diagnostics/profiles.json")) as Record<
		string,
		string[]
	>;
	const seenFeatures = new Set<string>();
	for (const file of files(path.join(root, "diagnostics")).filter((f) =>
		/\.mdx?$/.test(f),
	)) {
		const id = path
			.relative(path.join(root, "diagnostics"), file)
			.split(path.sep)
			.join("/")
			.replace(/\.mdx?$/, "");
		const feature = id.split("/")[0];
		seenFeatures.add(feature);
		if (!map[feature])
			throw new Error(`${id}: missing diagnostic profile mapping`);
		add("diagnostic", id, file, map[feature]);
	}
	if (!entries.some((c) => c.suite === "diagnostic"))
		throw new Error("diagnostic inventory is empty");
	for (const feature of Object.keys(map))
		if (!seenFeatures.has(feature))
			throw new Error(
				`${feature} (${map[feature]}): orphan diagnostic mapping`,
			);
	for (const p of PROFILES)
		if (!entries.some((c) => c.profile === p.name))
			throw new Error(`${p.name}: required suite is empty`);
	const expectedInputs = new Set(entries.map((c) => c.inputPath));
	for (const file of files(path.join(root, "corpus/inputs")))
		if (!expectedInputs.has(file)) throw new Error(`orphan input: ${file}`);
	const expected = new Set(entries.map((c) => c.goldenPath));
	for (const dir of ["goldens", "diagnostics-goldens"]) {
		if (!requireGoldens && !fs.existsSync(path.join(root, dir))) continue;
		for (const file of files(path.join(root, dir)))
			if (!expected.has(file))
				throw new Error(
					`orphan golden (profile/case): ${path.relative(root, file)}`,
				);
	}
	return entries;
}
export function inventoryCounts(cases: InventoryCase[]) {
	return Object.fromEntries(
		PROFILES.map((p) => [
			p.name,
			{
				corpus: cases.filter(
					(c) => c.profile === p.name && c.suite === "corpus",
				).length,
				diagnostic: cases.filter(
					(c) => c.profile === p.name && c.suite === "diagnostic",
				).length,
			},
		]),
	);
}

File: writr-rs/crates/writr-conformance/src/inventory.rs
use crate::{normalize, Golden, PROFILES};
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

fn json(path: &Path) -> Result<Value, String> {
	serde_json::from_str(
		&std::fs::read_to_string(path).map_err(|e| format!("{}: {e}", path.display()))?,
	)
	.map_err(|e| e.to_string())
}
fn files(dir: &Path) -> Result<Vec<PathBuf>, String> {
	let mut result = Vec::new();
	for e in std::fs::read_dir(dir).map_err(|e| format!("{}: {e}", dir.display()))? {
		let p = e.map_err(|e| e.to_string())?.path();
		if p.is_dir() {
			result.extend(files(&p)?);
		} else {
			result.push(p);
		}
	}
	result.sort();
	Ok(result)
}
fn add(
	out: &mut Vec<Golden>,
	root: &Path,
	id: &str,
	input: PathBuf,
	profiles: &Value,
	suite: &str,
) -> Result<(), String> {
	let profiles = profiles
		.as_array()
		.ok_or_else(|| format!("{id}: missing profile assignments"))?;
	if profiles.is_empty() {
		return Err(format!("{id}: empty profiles"));
	}
	for p in profiles {
		let name = p.as_str().ok_or("invalid profile")?;
		if crate::profile(name).is_none() {
			return Err(format!("{id}: unknown profile {name}"));
		}
		if out
			.iter()
			.any(|g| g.profile == name && g.id == id && g.input_path == input)
		{
			return Err(format!("{name} :: {id}: duplicate case"));
		}
		let golden = root.join(suite).join(name).join(format!("{id}.html"));
		if !input.is_file() || !golden.is_file() {
			return Err(format!("{name} :: {id}: missing input or golden"));
		}
		out.push(Golden {
			profile: name.into(),
			id: id.into(),
			input_path: input.clone(),
			golden_path: golden,
		});
	}
	Ok(())
}
/// Shared JSON inventories define every case. JS additionally verifies SHA-256.
pub fn discover(root: &Path) -> Result<Vec<Golden>, String> {
	let manifest = json(&root.join("corpus/manifest.json"))?;
	let entries = manifest["entries"]
		.as_array()
		.ok_or("missing manifest entries")?;
	if entries.is_empty() || manifest["count"].as_u64() != Some(entries.len() as u64) {
		return Err("manifest count/empty inventory".into());
	}
	let mut out = Vec::new();
	let mut ids = BTreeSet::new();
	let mut counts = BTreeMap::<String, usize>::new();
	for e in entries {
		let id = e["id"].as_str().ok_or("missing case id")?;
		if id.split('/').any(|x| x == ".." || x.is_empty())
			|| id.starts_with('/')
			|| id.contains('\\')
		{
			return Err(format!("invalid id {id}"));
		}
		if !ids.insert(id) {
			return Err(format!("duplicate case {id}"));
		}
		let expected_path = format!("{id}.md");
		if e["path"].as_str() != Some(&expected_path) {
			return Err(format!("{id}: input path differs"));
		}
		let input = root.join("corpus/inputs").join(expected_path);
		let raw = std::fs::read_to_string(&input).map_err(|error| format!("{id}: {error}"))?;
		if normalize(&raw) != raw || e["bytes"].as_u64() != Some(raw.len() as u64) {
			return Err(format!(
				"{} :: {id}: input bytes/normalization differ",
				e["profiles"]
			));
		}
		*counts
			.entry(e["source"].as_str().ok_or("missing source")?.into())
			.or_default() += 1;
		add(&mut out, root, id, input, &e["profiles"], "goldens")?;
	}
	if serde_json::to_value(counts).map_err(|e| e.to_string())? != manifest["bySource"] {
		return Err("manifest source counts differ".into());
	}
	let map = json(&root.join("diagnostics/profiles.json"))?;
	let mut features = BTreeSet::new();
	let mut diagnostic_ids = BTreeSet::new();
	for file in files(&root.join("diagnostics"))? {
		if !file.extension().is_some_and(|e| e == "md" || e == "mdx") {
			continue;
		}
		let id = file
			.strip_prefix(root.join("diagnostics"))
			.map_err(|e| e.to_string())?
			.with_extension("")
			.to_string_lossy()
			.replace('\\', "/");
		if !diagnostic_ids.insert(id.clone()) {
			return Err(format!("duplicate diagnostic {id}"));
		}
		let feature = id.split('/').next().ok_or("missing feature")?;
		features.insert(feature.to_owned());
		add(
			&mut out,
			root,
			&id,
			file,
			&map[feature],
			"diagnostics-goldens",
		)?;
	}
	if diagnostic_ids.is_empty() {
		return Err("empty diagnostic inventory".into());
	}
	for feature in map.as_object().ok_or("invalid diagnostic mapping")?.keys() {
		if !features.contains(feature) {
			return Err(format!("orphan diagnostic mapping {feature}"));
		}
	}
	for profile in PROFILES {
		if !out.iter().any(|g| g.profile == profile.name) {
			return Err(format!("{}: empty required suite", profile.name));
		}
	}
	let expected: BTreeSet<_> = out.iter().map(|g| g.golden_path.clone()).collect();
	for dir in ["goldens", "diagnostics-goldens"] {
		for file in files(&root.join(dir))? {
			if !expected.contains(&file) {
				return Err(format!("orphan golden {}", file.display()));
			}
		}
	}
	let expected_inputs: BTreeSet<_> = out.iter().map(|g| g.input_path.clone()).collect();
	for file in files(&root.join("corpus/inputs"))? {
		if !expected_inputs.contains(&file) {
			return Err(format!("orphan input {}", file.display()));
		}
	}
	Ok(out)
}

#[cfg(test)]
mod tests {
	use super::*;
	use serde_json::json;
	use std::sync::atomic::{AtomicUsize, Ordering};
	static NEXT: AtomicUsize = AtomicUsize::new(0);
	struct Fixture(PathBuf);
	impl Drop for Fixture {
		fn drop(&mut self) {
			let _ = std::fs::remove_dir_all(&self.0);
		}
	}
	impl Fixture {
		fn write(&self, path: &str, content: &str) {
			let path = self.0.join(path);
			std::fs::create_dir_all(path.parent().unwrap()).unwrap();
			std::fs::write(path, content).unwrap();
		}
		fn manifest(&self) -> Value {
			json(&self.0.join("corpus/manifest.json")).unwrap()
		}
		fn save(&self, manifest: &Value) {
			self.write("corpus/manifest.json", &manifest.to_string());
		}
	}
	fn fixture() -> Fixture {
		let dir = Fixture(std::env::temp_dir().join(format!(
			"writr-inventory-{}-{}",
			std::process::id(),
			NEXT.fetch_add(1, Ordering::Relaxed)
		)));
		std::fs::create_dir(&dir.0).unwrap();
		dir.write("corpus/inputs/test/one.md", "Hello\n");
		dir.save(&json!({"count":1,"bySource":{"test":1},"entries":[{"id":"test/one","path":"test/one.md","source":"test","bytes":6,"profiles":["default"]}]}));
		dir.write("goldens/default/test/one.html", "<p>Hello</p>\n");
		dir.write("diagnostics/text/one.md", "Hello\n");
		dir.write(
			"diagnostics/profiles.json",
			&json!({"text":PROFILES.iter().map(|p|p.name).collect::<Vec<_>>()}).to_string(),
		);
		for p in PROFILES {
			dir.write(
				&format!("diagnostics-goldens/{}/text/one.html", p.name),
				"<p>Hello</p>\n",
			);
		}
		dir
	}
	#[test]
	fn expected_cases_do_not_depend_on_golden_discovery() {
		let f = fixture();
		assert_eq!(discover(&f.0).unwrap().len(), 8);
		std::fs::remove_file(f.0.join("goldens/default/test/one.html")).unwrap();
		assert!(discover(&f.0)
			.unwrap_err()
			.contains("default :: test/one: missing input or golden"));
	}
	#[test]
	fn corrupt_inventories_fail_in_disposable_directories() {
		for scenario in [
			"manifest",
			"input",
			"bytes",
			"count",
			"source",
			"duplicate",
			"profile",
			"empty-profile",
			"path",
			"id",
			"orphan-golden",
			"orphan-input",
			"orphan-mapping",
			"duplicate-diagnostic",
			"mapping",
		] {
			let f = fixture();
			let mut m = f.manifest();
			match scenario {
				"manifest" => {
					std::fs::remove_file(f.0.join("corpus/manifest.json")).unwrap();
				}
				"input" => {
					std::fs::remove_file(f.0.join("corpus/inputs/test/one.md")).unwrap();
				}
				"bytes" => f.write("corpus/inputs/test/one.md", "Hello  \n"),
				"orphan-golden" => f.write("goldens/default/test/orphan.html", ""),
				"orphan-input" => f.write("corpus/inputs/test/orphan.md", ""),
				"orphan-mapping" => {
					let mut mapping = json(&f.0.join("diagnostics/profiles.json")).unwrap();
					mapping["orphan"] = json!(["default"]);
					f.write("diagnostics/profiles.json", &mapping.to_string());
				}
				"duplicate-diagnostic" => f.write("diagnostics/text/one.mdx", "Hello\n"),
				"mapping" => f.write("diagnostics/profiles.json", "{}"),
				_ => {
					match scenario {
						"count" => m["count"] = json!(2),
						"source" => m["bySource"]["test"] = json!(2),
						"duplicate" => {
							let entry = m["entries"][0].clone();
							m["entries"].as_array_mut().unwrap().push(entry);
							m["count"] = json!(2);
						}
						"profile" => m["entries"][0]["profiles"] = json!(["unknown"]),
						"empty-profile" => m["entries"][0]["profiles"] = json!([]),
						"path" => m["entries"][0]["path"] = json!("elsewhere.md"),
						"id" => m["entries"][0]["id"] = json!("../outside"),
						_ => unreachable!(),
					}
					f.save(&m);
				}
			}
			assert!(discover(&f.0).is_err(), "{scenario}");
		}
	}
}
